extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java

```diff
@@ -212,41 +213,193 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
+        BlockingMode mode = BlockingMode.UNDEFINED;
+        for (ClassInfo ci : classes) {
+            MethodInfo method = ci.method(methodName, methodArgs);
+            if (method != null) {
+                Predicate<DotName> annotationOnMethod = n -> {
+                    AnnotationInstance annotationInstance = method.annotation(n);
+                    return annotationInstance != null && annotationInstance.target().kind() == Kind.METHOD;
+                };
+                BlockingMode methodMode = blockingMode(annotationOnMethod,
+                        () -> "Method '" + method.declaringClass().name() + "#" + method.name() +
+                                "' contains both @Blocking and @NonBlocking annotations.");
+                if (methodMode.explicit) {
+                    return methodMode;
                 }
-            } else if ((method.hasAnnotation(TRANSACTIONAL) || blockingClass) && !method.hasAnnotation(NON_BLOCKING)) {
-                result.add(method);
+                if (methodMode != BlockingMode.UNDEFINED) {
+                    mode = methodMode;
+                }
+            }
+        }
+        return mode;
+    }
+
+    private boolean effectiveBlockingMode(BlockingMode methodBlockingMode, BlockingMode serviceBlockingMode) {
+        if (methodBlockingMode.explicit) {
+            // explicit @Blocking or @NonBlocking on (any) method
+            return methodBlockingMode.blocking;
+        }
+        if (serviceBlockingMode.explicit) {
+            // explicit @Blocking or @NonBlocking on (any) class
+            return serviceBlockingMode.blocking;
+        }
+        // @Transactional on method or class
+        return methodBlockingMode.blocking || serviceBlockingMode.blocking;
+    }
+
+    /**
+     * Collect the names of all blocking methods.
+     *
+     * <p>
+     * Whether a method is blocking or not is evaluated for each individual service method (those that are defined
+     * in the generated {@code *ImplBase} class).
+     *
+     * <p>
+     * For each method:
+     * <ol>
+     * <li>blocking, if any of the method overrides has a {@link io.smallrye.common.annotation.Blocking} annotation.</li>
+     * <li>not-blocking, if any of the method overrides has a {@link io.smallrye.common.annotation.NonBlocking} annotation.</li>
+     * <li>blocking, if the service class or any of its base classes has a {@link io.smallrye.common.annotation.Blocking}
+     * annotation.</li>
+     * <li>non-blocking, if the service class or any of its base classes has a {@link io.smallrye.common.annotation.NonBlocking}
+     * annotation.</li>
+     * <li>blocking, if any of the method overrides has a {@link javax.transaction.Transaction} annotation.</li>
+     * <li>blocking, if the service class or any of its base classes has a {@link javax.transaction.Transaction}
+     * annotation.</li>
+     * <li>Else: non-blocking.</li>
```

Comment:
As I wrote before, I think this is how it should work:
- apply method-level @Blocking, @NonBlocking and blocking coming from @Transactional only to methods that define it (and not on overriding methods)
- apply class-level annotations only to methods in the class they are in

Even with this simplification, we're ending up with quite strange cases like:
```java
@Blocking
class BaseClass {
   void foo(){}
}

@NonBlocking
class SubClass extends BaseClass {
 @Override
 void foo() {
 }
}


class SubClass2 extends BaseClass {
 @Override
 void foo() {
 }
}
```

or 
```java
class BaseClass {
  @Blocking
   void foo(){}
}

@NonBlocking
class SubClass extends BaseClass {
}
```

Reply:
To my knowledge we don't support inheriting `@Blocking` in RestEasy Reactive (yet?).

@Ladicek, maybe you have some opinion on how this would work best?

Reply:
I certainly have strong opinions on annotation inheritance (it's evil!), but I also know how annotations are inherited in the Java language: they are never inherited from methods, only from superclasses. The Common Annotations specification demands the same, and it's also how CDI behaves. Interestingly, JAX-RS behaves differently, which is probably the source of inspiration here.

Here's my take: the `@Blocking` and `@NonBlocking` annotations are _not_ declared as `@Inherited`. Don't inherit them. Anywhere. Ever.

Reply:
Yea, it's been inspired by JAX-RS. But I get the (strong) opinion and agree that this implementation should conform to the documentation of `@Blocking` and `@NonBlocking`. `@Transactional` however is `@Inherited`, which makes it a bit tricky.

Anyway, I've updated the implementation and added a unit test for a bunch of combinations to check the blocking methods for a particular class.

Reply:
is there any test that would check if in the following cases the `foo` method is non-blocking?
```java
@Blocking 
class BaseClass { 
   void foo(){} 
} 
class SubClass extends BaseClass { @Override void foo() { } 
} 
```

```java
class BaseClass { 
   @Blocking 
   void foo(){} 
} 
class SubClass extends BaseClass { 
   @Override void foo() { } 
} 
```

Reply:
also, what's the reason for including every kind of communication type (unary, client and server side streaming, bidirectional) in the tests for this PR? Isn't it enough to test only one case, e.g. unary?

Reply:
I followed what the existing tests were doing to check the blocking behavior so I'm not sure actually.

Re unit test cases: added a couple more following the pattern you asked for.